plugdj/events.py

```diff
@@ -1,33 +1,49 @@
-from sys import exc_info
 from .util import MalformedEvent
 
 def from_json(js):
-    ev_class = event_map.get(js.get("a"))
-    return UnknownEvent(js) if ev_class is None else ev_class(js.get("p"))
+    EventClass = event_map.get(js.get("a"))
+    return EventClass(js) if EventClass is not None else UnknownEvent(js)
 
 # because namedtuple is too restrictive; ignore extras
 class PlugEvent(object):
     __slots__ = ()
+
     def __init__(self, json):
         for attr in self.__slots__:
             try:
-                setattr(self, attr, json[attr])
-            except KeyError:
-                _, _, trace = exc_info()
+                p = json["p"]
+                setattr(self, attr, p[attr] if hasattr(p, "__getitem__") else p)
```

Comment:
This looks a bit weird. Would you mind explaining a bit here? Thanks in advance.

Reply:
I feel I must have encountered a {"p": <non-map-type>} i.e. not a dict. For now, reverted.

Reply:
Oh, I remember the example now... 

```json
'[{"a":"friendRequest","p":"nicfit","s":"dashboard"}]
```

'p' is not a dict, but is is valid json. I guess I'll bring this code back.